File: C/chapitre5/Exercice9.c
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <semaphore.h>
#include <unistd.h>
#include <time.h>

sem_t plein;
sem_t vide;
unsigned char buf;

void *emetteur(void *arg) {
    while (1) {
        buf = rand() % 256;
        sem_post(&plein);
        sem_wait(&vide);
        sleep(1 + rand() % 3);
    }
    return NULL;
}

void *recepteur(void *arg) {
    while (1) {
        sem_wait(&plein);
        printf("Lu: %d\n", buf);
        sem_post(&vide);
        sleep(1 + rand() % 3);
    }
    return NULL;
}

int main(void) {
    sem_init(&plein, 0, 0);
    sem_init(&vide, 0, 1);
    srand(time(NULL));
    pthread_t e;
    pthread_t r;
    pthread_create(&e, NULL, emetteur, NULL);
    pthread_create(&r, NULL, recepteur, NULL);
    pthread_join(e, NULL);
    pthread_join(r, NULL);
    sem_destroy(&plein);
    sem_destroy(&vide);
    return 0;
}

File: Python/chapitre5/Exercice9.py
import threading
import time
import random

plein = threading.Semaphore(0)
vide = threading.Semaphore(1)
buf = [0]

def emetteur():
    while True:
        octet = random.randint(0, 255)
        vide.acquire()
        buf[0] = octet
        plein.release()
        time.sleep(random.randint(1, 3))

def recepteur():
    while True:
        plein.acquire()
        val = buf[0]
        print("Lu:", val)
        vide.release()
        time.sleep(random.randint(1, 3))

if __name__ == "__main__":
    te = threading.Thread(target=emetteur)
    tr = threading.Thread(target=recepteur)
    te.start()
    tr.start()
    te.join()
    tr.join()